src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,504 @@
+            {
+                HttpResponseMessage response;
+                try
+                {
+                    response = await __httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+                }
+                catch (Exception ex) when (ex is OperationCanceledException || ex is MongoClientException)
+                {
+                    throw new MongoClientException(exceptionMessage, ex);
+                }
+
+                return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            }
+        }
+
+        private class MongoAWSMechanism : ISaslMechanism
+        {
+            private readonly IClock _clock;
+            private readonly AwsCredentials _awsCredentials;
+            private readonly IRandomByteGenerator _randomByteGenerator;
+
+            public MongoAWSMechanism(
+                AwsCredentials awsCredentials,
+                IRandomByteGenerator randomByteGenerator,
+                IClock clock)
+            {
+                _awsCredentials = Ensure.IsNotNull(awsCredentials, nameof(awsCredentials));
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+                _clock = Ensure.IsNotNull(clock, nameof(clock));
+            }
+
+            public string Name
+            {
+                get { return MechanismName; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument
+                {
+                    { "r", nonce },
+                    { "p", (int)'n' }
+                };
+
+                var clientMessageBytes = document.ToBson();
+
+                return new ClientFirst(clientMessageBytes, nonce, _awsCredentials, _clock);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(ClientNonceLength);
+            }
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly IClock _clock;
+            private readonly AwsCredentials _awsCredentials;
+            private readonly byte[] _nonce;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                AwsCredentials awsCredentials,
+                IClock clock)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _clock = clock;
+                _awsCredentials = awsCredentials;
+                _nonce = nonce;
```

Comment:
Normally we process the arguments in the order they are declared in the method:

    _bytesToSendToServer = bytesToSendToServer;
    _nonce = nonce;
    _awsCredentials = awsCredentials;
    _clock = clock;